src/nvim/os/env.c

```diff
@@ -2,10 +2,20 @@
 
 #include <uv.h>
 
+// vim.h must be included before charset.h (and possibly others) or things
+// blow up
```

Comment:
Yep, `charset.h.generated.h` depends on a few types being defined before it's included: `char_u`, `buf_T`, `colnr_T`, and `win_T`. 

This might be a build system bug (assuming every header file should "just work" without depending on ordering) but that's probably for another PR.